tidy3d/plugins/mode/mode_solver.py

```diff
@@ -348,6 +366,422 @@ def data_raw(self) -> ModeSolverData:
+            "f": f,
+            "mode_index": mode_index,
+        }
+
+        solver_data_cylindrical = {
+            name: ScalarModeFieldCylindricalDataArray(field_data_cylindrical[name], coords=coords)
+            for name in ("Er", "Etheta", "Eaxial", "Hr", "Htheta", "Haxial")
+        }
+        solver_data_cylindrical["n_complex"] = mode_solver_data.n_complex
+
+        return solver_data_cylindrical
+
+    def _mode_rotation(
+        self,
+        solver_ref_data_cylindrical: Dict[
+            Union[ScalarModeFieldCylindricalDataArray, ModeIndexDataArray]
+        ],
+        solver: ModeSolver,
+    ) -> ModeSolverData:
+        """Rotate the mode solver solution from the reference plane in cylindrical coordinates
+        to the desired monitor plane."""
+        rotated_data_arrays = {}
+        for field_name in ("Ex", "Ey", "Ez", "Hx", "Hy", "Hz"):
+            if self.colocate is True:
+                # Get colocation coordinates in the solver plane
+                normal_dim, _ = self.plane.pop_axis("xyz", self.normal_axis)
+                colocate_coords = self._get_colocation_coordinates()
+                colocate_coords[normal_dim] = np.atleast_1d(self.plane.center[self.normal_axis])
+                x = colocate_coords["x"]
+                y = colocate_coords["y"]
+                z = colocate_coords["z"]
+                xyz_coords = [x.copy(), y.copy(), z.copy()]
+            else:
+                # Extract coordinate values from one of the six field components
+                xyz_coords = solver.grid_snapped[field_name].to_list
+                x, y, z = (coord.copy() for coord in xyz_coords)
+
+            f = np.atleast_1d(self.freqs)
+            mode_index = np.arange(self.mode_spec.num_modes)
+
+            axial = solver_ref_data_cylindrical["Er"].axial
+
+            # Initialize output arrays
+            shape = (x.size, y.size, z.size, len(f), mode_index.size)
+            rotated_field_cmp = np.zeros(shape, dtype=complex)
+
+            idx_w, idx_uv = self.plane.pop_axis((0, 1, 2), axis=self.bend_axis_3d)
+            idx_u, idx_v = idx_uv
+
+            pts = [coord.copy() for coord in xyz_coords]
+
+            pts[idx_u] -= self.bend_center[idx_u]
+            pts[idx_v] -= self.bend_center[idx_v]
+
+            rho = np.sqrt(pts[idx_u] ** 2 + pts[idx_v] ** 2)
+            theta = np.arctan2(pts[idx_v], pts[idx_u])
+            axial = pts[idx_w]
+
+            theta_rel = theta - self.theta_reference
+
+            cos_theta = pts[idx_u] / rho
+            sin_theta = pts[idx_v] / rho
+
+            cmp_normal, source_names = self.plane.pop_axis(("x", "y", "z"), axis=self.bend_axis_3d)
+            cmp_1, cmp_2 = source_names
+
+            k0 = 2 * np.pi * f / C_0
+            n_eff = solver_ref_data_cylindrical["n_complex"].values
+            beta = k0[:, None] * n_eff * np.abs(self.mode_spec.bend_radius)
+
+            for axial_idx, axial_value in enumerate(axial):
```

Comment:
Seems like this should be easily vectorizable too?

Reply:
Yeah, I've updated it with vectorization, and now it's 80% faster compared to when `bend_angle_rotation=False`. Before, it was slower.